File: ios/ChronosCore/Sources/ChronosCore/FoldedReminder.swift
import Foundation

/// The iOS mirror of Android `FoldedReminder.kt` — pure ranking and per-kind due rules for the
/// schedule Live Activity fold. Keeps the coordinator thin and unit-testable without SwiftData.
public enum FoldedReminderKind: Int, Hashable, Sendable {
    case medication = 0
    case task = 1
    case habit = 2
}

public struct FoldedReminder: Hashable, Sendable {
    public var kind: FoldedReminderKind
    public var entityID: String
    public var title: String
    public var detail: String
    public var dueMinute: Int
    public var isOverdue: Bool

    public init(
        kind: FoldedReminderKind,
        entityID: String,
        title: String,
        detail: String,
        dueMinute: Int,
        isOverdue: Bool
    ) {
        self.kind = kind
        self.entityID = entityID
        self.title = title
        self.detail = detail
        self.dueMinute = dueMinute
        self.isOverdue = isOverdue
    }
}

/// Matches Android `OVERDUE_THRESHOLD_MINUTES` and iOS reminder snooze cadence.
public let foldedReminderOverdueThresholdMinutes = 15

/// Matches Android `MAX_FOLDED_REMINDERS`.
public let maxFoldedReminders = 3

public func rankFoldedReminders(
    _ candidates: [FoldedReminder],
    limit: Int = maxFoldedReminders
) -> [FoldedReminder] {
    candidates
        .sorted {
            let ka = foldedReminderKindWeight($0.kind), kb = foldedReminderKindWeight($1.kind)
            if ka != kb { return ka < kb }
            return $0.dueMinute < $1.dueMinute
        }
        .prefix(Swift.max(0, limit))
        .map { $0 }
}

private func foldedReminderKindWeight(_ kind: FoldedReminderKind) -> Int {
    kind.rawValue
}

// MARK: - Medication

public struct MedicationFoldInput: Sendable {
    public var id: String
    public var name: String
    public var dosage: String
    public var unit: String
    public var isActive: Bool
    public var reminderMinutes: [Int]
    public var pausedUntil: Date?
    public var takenScheduledMinutes: Set<Int>

    public init(
        id: String,
        name: String,
        dosage: String,
        unit: String,
        isActive: Bool,
        reminderMinutes: [Int],
        pausedUntil: Date?,
        takenScheduledMinutes: Set<Int>
    ) {
        self.id = id
        self.name = name
        self.dosage = dosage
        self.unit = unit
        self.isActive = isActive
        self.reminderMinutes = reminderMinutes
        self.pausedUntil = pausedUntil
        self.takenScheduledMinutes = takenScheduledMinutes
    }
}

public func buildMedicationFoldedReminders(
    plans: [MedicationFoldInput],
    today: Date,
    nowMinute: Int,
    calendar: Calendar = .current
) -> [FoldedReminder] {
    let startToday = calendar.startOfDay(for: today)
    return plans.flatMap { plan -> [FoldedReminder] in
        guard plan.isActive else { return [] }
        if let paused = plan.pausedUntil, startToday < calendar.startOfDay(for: paused) { return [] }
        guard let due = plan.reminderMinutes.filter({ $0 <= nowMinute }).max() else { return [] }
        guard !plan.takenScheduledMinutes.contains(due) else { return [] }
        let dose = "\(plan.dosage) \(plan.unit)".trimmingCharacters(in: .whitespaces)
        let detail = dose.isEmpty ? "Due \(clockTime(due))" : "Due \(clockTime(due)) · \(dose)"
        return [
            FoldedReminder(
                kind: .medication,
                entityID: plan.id,
                title: plan.name,
                detail: detail,
                dueMinute: due,
                isOverdue: nowMinute - due >= foldedReminderOverdueThresholdMinutes)
        ]
    }
}

// MARK: - Tasks

public struct TaskFoldInput: Sendable {
    public var id: String
    public var title: String
    public var isCompleted: Bool
    public var dueDate: Date?
    public var priority: Int

    public init(
        id: String,
        title: String,
        isCompleted: Bool,
        dueDate: Date?,
        priority: Int
    ) {
        self.id = id
        self.title = title
        self.isCompleted = isCompleted
        self.dueDate = dueDate
        self.priority = priority
    }
}

public func buildTaskFoldedReminders(
    tasks: [TaskFoldInput],
    today: Date,
    nowMinute: Int,
    calendar: Calendar = .current
) -> [FoldedReminder] {
    tasks.compactMap { task in
        guard !task.isCompleted, let due = task.dueDate, calendar.isDate(due, inSameDayAs: today) else {
            return nil
        }
        let dueMin = calendar.component(.hour, from: due) * 60 + calendar.component(.minute, from: due)
        guard dueMin <= nowMinute else { return nil }
        let detail = task.priority >= 3
            ? "Due \(clockTime(dueMin)) · High priority"
            : "Due \(clockTime(dueMin))"
        return FoldedReminder(
            kind: .task,
            entityID: task.id,
            title: task.title,
            detail: detail,
            dueMinute: dueMin,
            isOverdue: nowMinute - dueMin >= foldedReminderOverdueThresholdMinutes)
    }
}

// MARK: - Habits

public struct HabitFoldInput: Sendable {
    public var id: String
    public var title: String
    public var isActive: Bool
    public var isDueToday: Bool
    public var isCompletedToday: Bool
    public var isSkippedToday: Bool
    public var isPausedToday: Bool
    public var windowOpenMinute: Int
    public var streakCount: Int

    public init(
        id: String,
        title: String,
        isActive: Bool,
        isDueToday: Bool,
        isCompletedToday: Bool,
        isSkippedToday: Bool,
        isPausedToday: Bool,
        windowOpenMinute: Int,
        streakCount: Int
    ) {
        self.id = id
        self.title = title
        self.isActive = isActive
        self.isDueToday = isDueToday
        self.isCompletedToday = isCompletedToday
        self.isSkippedToday = isSkippedToday
        self.isPausedToday = isPausedToday
        self.windowOpenMinute = windowOpenMinute
        self.streakCount = streakCount
    }
}

public func buildHabitFoldedReminders(
    habits: [HabitFoldInput],
    nowMinute: Int
) -> [FoldedReminder] {
    habits.compactMap { habit in
        guard habit.isActive, habit.isDueToday,
              !habit.isCompletedToday, !habit.isSkippedToday, !habit.isPausedToday
        else { return nil }
        let open = habit.windowOpenMinute
        guard open <= nowMinute else { return nil }
        let detail = habit.streakCount > 0
            ? "Window open · \(habit.streakCount)-day streak"
            : "Window open"
        return FoldedReminder(
            kind: .habit,
            entityID: habit.id,
            title: habit.title,
            detail: detail,
            dueMinute: open,
            isOverdue: nowMinute - open >= foldedReminderOverdueThresholdMinutes)
    }
}

File: core/notifications/src/main/java/com/ChronosFlow/VBCR/core/notifications/FoldedReminder.kt
package com.ChronosFlow.VBCR.core.notifications

import com.ChronosFlow.VBCR.core.domain.model.Habit
import com.ChronosFlow.VBCR.core.domain.model.MedicationDoseEventType
import com.ChronosFlow.VBCR.core.domain.model.MedicationPlan
import com.ChronosFlow.VBCR.core.domain.model.Task
import com.ChronosFlow.VBCR.core.domain.model.buildLegacyHabitSchedule
import java.time.LocalDate
import java.time.ZoneId

/**
 * The "folded reminders" fold — the Android mirror of iOS `BlockLiveActivityCoordinator.foldedReminders`.
 * When the current-block ("now" / up-next) live notification is showing, the single most-imminent DUE
 * reminder (a medication dose, a task, or a habit window) is surfaced ON that surface as a glanceable text
 * segment plus one action button. When fold is enabled, [AlarmDeliveryCoordinator] suppresses
 * separate med/task/habit banners for items already on this surface (iOS passive-demotion parity).
 *
 * Everything in this file is pure and Android-free so the ranking and per-kind due rules are unit-testable
 * without a device — the repository plumbing lives in [FoldedReminderResolver].
 */

/** Which entity a folded reminder came from — drives ranking priority and the action-button label/receiver. */
enum class FoldedReminderKind { MEDICATION, TASK, HABIT }

/**
 * A reminder that is already actionable today. [dueMinute] is the minute-of-day it became due (for ranking
 * the most-overdue first); [isOverdue] is true once at least [OVERDUE_THRESHOLD_MINUTES] have passed.
 */
data class FoldedReminder(
    val kind: FoldedReminderKind,
    val entityId: String,
    val title: String,
    val detail: String,
    val dueMinute: Int,
    val isOverdue: Boolean
)

/**
 * A dose/task/habit reads as "overdue" this many minutes after its time passes — matches the reminder
 * snooze interval so a just-due item still reads as "due now" first (mirrors iOS `overdueThreshold`).
 */
internal const val OVERDUE_THRESHOLD_MINUTES = 15

/** How many folded reminders the live surface carries at once before collapsing the rest into "＋N more". */
internal const val MAX_FOLDED_REMINDERS = 3

/**
 * Ranks folded reminders the way the live surface presents them: medication first (health-critical), then
 * task, then habit; within a kind the most-overdue (earliest [FoldedReminder.dueMinute]) first. Capped at
 * [max]. The first element is the primary chip / action.
 */
internal fun rankFoldedReminders(
    candidates: List<FoldedReminder>,
    max: Int = MAX_FOLDED_REMINDERS
): List<FoldedReminder> =
    candidates
        .sortedWith(compareBy({ kindWeight(it.kind) }, { it.dueMinute }))
        .take(max.coerceAtLeast(0))

/** Entity identity for fold suppression — mirrors iOS `FoldedReminderLiveResolver.EntityKey`. */
internal data class FoldedEntityKey(val kind: FoldedReminderKind, val entityId: String)

internal fun foldedEntityKeys(reminders: List<FoldedReminder>): Set<FoldedEntityKey> =
    reminders.map { FoldedEntityKey(it.kind, it.entityId) }.toSet()

/**
 * When fold is on, skip separate med/task/habit banners already surfaced on the live "now"
 * notification — mirrors iOS `ChronosNotifications.shouldSuppressFoldedReminderBanner`.
 */
internal fun shouldSuppressFoldedReminderBanner(
    folded: Set<FoldedEntityKey>,
    medicationPlanId: String?,
    taskId: String?,
    habitId: String?
): Boolean {
    if (folded.isEmpty()) return false
    medicationPlanId?.let { id ->
        if (folded.contains(FoldedEntityKey(FoldedReminderKind.MEDICATION, id))) return true
    }
    taskId?.let { id ->
        if (folded.contains(FoldedEntityKey(FoldedReminderKind.TASK, id))) return true
    }
    habitId?.let { id ->
        if (folded.contains(FoldedEntityKey(FoldedReminderKind.HABIT, id))) return true
    }
    return false
}

/** Ranking weight: medication 0 < task 1 < habit 2 (mirrors iOS `kindWeight`). */
private fun kindWeight(kind: FoldedReminderKind): Int = when (kind) {
    FoldedReminderKind.MEDICATION -> 0
    FoldedReminderKind.TASK -> 1
    FoldedReminderKind.HABIT -> 2
}

/** All reminder minutes for a plan — schedule times when present, else the legacy single minute. */
internal fun medicationReminderMinutes(plan: MedicationPlan): List<Int> {
    val scheduled = plan.schedule?.recurrence?.timesOfDayMinutes
        ?.map { ((it % 1440) + 1440) % 1440 }
        ?.filter { it in 0..1439 }
        ?.distinct()
        ?.sorted()
    return if (!scheduled.isNullOrEmpty()) scheduled else listOf(plan.reminderMinuteOfDay)
}

/** Whether reminders are suppressed because the plan is paused through [date] (inclusive). */
internal fun MedicationPlan.isPaused(date: LocalDate = LocalDate.now()): Boolean =
    schedule?.pausedUntil?.let { !it.isBefore(date) } == true

/** True when a TAKEN event covers the scheduled dose at [minute] on [date]. */
internal fun MedicationPlan.isDoseTaken(date: LocalDate, minute: Int): Boolean =
    recentDoseEvents.any {
        it.eventDate == date &&
            it.type == MedicationDoseEventType.TAKEN &&
            (it.scheduledMinuteOfDay == null || it.scheduledMinuteOfDay == minute)
    }

/**
 * Medication doses due right now: plan active, not paused, and the latest passed reminder minute
 * for today has not been taken. Supports multi-dose schedules via [medicationReminderMinutes].
 * Mirrors iOS `medicationReminders`.
 */
internal fun buildMedicationFoldedReminders(
    plans: List<MedicationPlan>,
    today: LocalDate,
    nowMinute: Int
): List<FoldedReminder> = plans.flatMap { plan ->
    if (!plan.isActive || plan.isPaused(today)) return@flatMap emptyList()
    val due = medicationReminderMinutes(plan).filter { it <= nowMinute }.maxOrNull()
        ?: return@flatMap emptyList()
    if (plan.isDoseTaken(today, due)) return@flatMap emptyList()
    val dose = "${plan.dosage} ${plan.unit}".trim()
    val detail = if (dose.isBlank()) {
        "Due ${formatCurrentBlockMinute(due)}"
    } else {
        "Due ${formatCurrentBlockMinute(due)} · $dose"
    }
    listOf(
        FoldedReminder(
            kind = FoldedReminderKind.MEDICATION,
            entityId = plan.id,
            title = plan.name,
            detail = detail,
            dueMinute = due,
            isOverdue = nowMinute - due >= OVERDUE_THRESHOLD_MINUTES
        )
    )
}

/**
 * Tasks due today whose due time has passed: not completed, [Task.dueDate] falls on [today] (in [zoneId]),
 * and its minute-of-day `<= nowMinute`. Detail reads "Due 9:00 AM", or "Due 9:00 AM · High priority" when
 * `priority >= 3`. Mirrors iOS `taskReminders`.
 */
internal fun buildTaskFoldedReminders(
    tasks: List<Task>,
    today: LocalDate,
    nowMinute: Int,
    zoneId: ZoneId
): List<FoldedReminder> = tasks.mapNotNull { task ->
    if (task.isCompleted) return@mapNotNull null
    val due = task.dueDate ?: return@mapNotNull null
    val dueLocal = due.atZone(zoneId).toLocalDateTime()
    if (dueLocal.toLocalDate() != today) return@mapNotNull null
    val dueMinute = dueLocal.hour * 60 + dueLocal.minute
    if (dueMinute > nowMinute) return@mapNotNull null
    val detail = if (task.priority >= 3) {
        "Due ${formatCurrentBlockMinute(dueMinute)} · High priority"
    } else {
        "Due ${formatCurrentBlockMinute(dueMinute)}"
    }
    FoldedReminder(
        kind = FoldedReminderKind.TASK,
        entityId = task.id,
        title = task.title,
        detail = detail,
        dueMinute = dueMinute,
        isOverdue = nowMinute - dueMinute >= OVERDUE_THRESHOLD_MINUTES
    )
}

/**
 * Habits whose window is open now: active, due today (recurrence lands and not completed/skipped/paused via
 * the shared [isReminderDueOnDate] rules), and the window (deferral, else start) has opened
 * (`open <= nowMinute`). Detail reads "Window open", or "Window open · 5-day streak" when a streak is running.
 * Mirrors iOS `habitReminders`; reuses [HabitDueRules] so the fold and the alarm scheduler never drift.
 */
internal fun buildHabitFoldedReminders(
    habits: List<Habit>,
    today: LocalDate,
    nowMinute: Int
): List<FoldedReminder> = habits.mapNotNull { habit ->
    if (!habit.isActive) return@mapNotNull null
    if (!habit.isReminderDueOnDate(today, habit.recurrenceReferenceDate(today))) return@mapNotNull null
    val schedule = habit.schedule ?: buildLegacyHabitSchedule(
        habitId = habit.id,
        cadence = habit.cadence,
        windowStartMinute = habit.windowStartMinute,
        windowEndMinute = habit.windowEndMinute,
        plannerVisible = habit.isBundled
    )
    // Honour a one-off deferral ("remind me later"); otherwise the window start.
    val open = schedule.deferUntilMinuteOfDay ?: schedule.targetStartMinute
    if (open > nowMinute) return@mapNotNull null
    val detail = if (habit.streakCount > 0) {
        "Window open · ${habit.streakCount}-day streak"
    } else {
        "Window open"
    }
    FoldedReminder(
        kind = FoldedReminderKind.HABIT,
        entityId = habit.id,
        title = habit.title,
        detail = detail,
        dueMinute = open,
        isOverdue = nowMinute - open >= OVERDUE_THRESHOLD_MINUTES
    )
}
